### PR TITLE
docs(dependency-injection): revise Dart and TS code and prose

### PR DESCRIPTION
**Pull request by chalin**
Contributes to #1508 and #1522.

- The main change on the TS side is to present best practices first, rather than presenting sample code and then saying "don't do that".
- The "Optional dependencies" section has been moved towards the end of the provider variants section and its sample code has been simplified.
- Reworked *_1.{ts,dart} sources so that main_1.{ts,dart} can actually be run, which helped find some bugs in those sources.
- Despite https://github.com/angular/angular/issues/8928, the Dart and TS OpaqueToken sections have been left in the prose. They can be revisited later.

Enhancements to the Dart code include:

- Added TestComponent.
- Added a provider sample that makes use of an app config Map (to match TS); also added the more Dart-like use of a custom app configuration class, whose instance is initialized via the cascade operator.
- Fixed Engine class hierarchy so that a single field named `cylinders` is defined (rather than two fields: one in the Engine supertype and one in the subclass), because having two fields is a bad design smell.
- Similarly, fixed Logger hierarchy so that a single field named `logs` is defined. Also made `logs` a read-only field.
- Reintroduced the SilentLogger().

### CODE REVIEW

**Comment by chalin** [comment]
@Foxandxss @wardbell @thso @kwalrath : this is ready for review. As usual, it is less painful to view the [diff while ignoring changes in whitespace (such as indentation)](https://github.com/angular/angular.io/pull/1573/files?w=1).

File: public/docs/_examples/dependency-injection/dart/lib/app_component.dart [hunk 1]
```diff
@@ -11,6 +11,7 @@ import 'user_service.dart';
 //PENDING: check whether we intend to hide injector_component.dart & providers_component.dart; if so, change docregion name?
```
Comment (kwalrath) [inline public/docs/_examples/dependency-injection/dart/lib/app_component.dart]: Should this comment really be visible? Also, it doesn't reflect test_component.dart.

File: public/docs/_examples/dependency-injection/dart/lib/app_component.dart [hunk 2]
```diff
@@ -31,21 +32,21 @@ import 'providers_component.dart';
...
     ],
-// #docregion providers
+    // #docregion providers
     providers: const [
-      Logger,
-      UserService,
-      const Provider(AppConfig, useValue: config1)]
-// #enddocregion providers
+      Logger, UserService,
+      const Provider(APP_CONFIG, useFactory: heroDiConfigFactory)]
```
Comment (kwalrath) [inline public/docs/_examples/dependency-injection/dart/lib/app_component.dart]: APP_CONFIG seems like an un-Darty name. (But if it's mentioned in text, it'd be easier to leave this capitalization.)

File: public/docs/_examples/dependency-injection/dart/lib/app_component.dart (hunk 1 above)
Comment (chalin) [inline public/docs/_examples/dependency-injection/dart/lib/app_component.dart]: That was an old comment that I missed (I had only searched for todo|fixme). I have removed it since it is not relevant (the entire source file is never shown).

File: public/docs/_examples/dependency-injection/dart/lib/app_component.dart (hunk 2 above)
Comment (chalin) [inline public/docs/_examples/dependency-injection/dart/lib/app_component.dart]: The OpaqueToken section will need to be revisited once #1563 gets resolved so I decided to leave the token name like it is in the TS example (because then the token name _and_ the interface name will match, both being `AppConfig`).

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 3]
```diff
@@ -9,34 +10,35 @@ include ../_util-fns
...
 :marked
   Our `Car` creates everything it needs inside its constructor.
   What's the problem?
-
   The problem is that our `Car` class is brittle, inflexible, and hard to test.
 
   Our `Car` needs an engine and tires. Instead of asking for them,
-  the `Car` constructor creates its own copies by "new-ing" them from
+  the `Car` constructor instantiates its own copies from
   the very specific classes, `Engine` and `Tires`.
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Remove the "," after "classes" while you're in here. (It's a restrictive phrase, not a parenthetical one.) _Or_ delete the "the" before "very" and keep the comma; that way the phrase becomes parenthetical.

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 4]
```diff
@@ -377,144 +375,110 @@ include ../_util-fns
...
+      li <b>Consistency:</b> All services follow the same rules, and we don't have to wonder why #{_a} #{_decorator} is missing.
+
+:marked
+  Injectors are also responsible for instantiating components
+  like `HeroesComponent`. Why haven't we marked `HeroesComponent` as
+  `@Injectable()`?
+
+  We *can* add it if we really want to. It isn't necessary because the
+  `HeroesComponent` is already marked with `@Component`, and this
+  !{_decorator} class (like `@Directive` and `@Pipe` which we will study later
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: "study" sounds so serious. Also, "which" (when introducing a phrase/clause) should have a comma before it; if a comma doesn't seem right, then you should probably be using "that" instead.

```
` which we will study later on
```
->
```
`, which we'll see later
```

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 5]
```diff
@@ -377,144 +375,110 @@ include ../_util-fns
...
+  on) is a subtype of <a href="#{injMetaUrl}">InjectableMetadata</a>.  It is in
+  fact <a href="#{injMetaUrl}">InjectableMetadata</a> #{_decorator}s that
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Remove the link from the second occurrence of InjectableMetadata.

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 3 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: removed ","

File: public/docs/_examples/dependency-injection/dart/lib/logger_service.dart [hunk 6]
```diff
@@ -3,10 +3,12 @@ import 'package:angular2/core.dart';
 
 @Injectable()
 class Logger {
-  List<String> logs = [];
+  List<String> _logs = [];
+
```
Comment (kwalrath) [inline public/docs/_examples/dependency-injection/dart/lib/logger_service.dart]: The blank line above the getter seems excessive. Is it standard? (It could be; I don't happen to know, and I don't see anything definitive in Effective Dart.)

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 4 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Fixed
> (like `@Directive` and `@Pipe`, which we'll learn about later)

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 5 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: done

**Comment by Foxandxss** [comment]
I have nothing to add, @kwalrath is doing an awesome job on the review. Sounds fine for me.

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 7]
```diff
@@ -748,105 +717,117 @@ code-example(format, language="html").
...
+  +makeExcerpt('app/app.component.ts','providers')
+
+#optional
 :marked
-// end typescript only
+  ## Optional dependencies
+
+  Our `HeroService` *requires* a `Logger`, but what if it could get by without
+  a logger?
+  As is illustrated next, we can indicate that to the dependency injection
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: This sounds very formal. Also, we should mention `@Optional()` in the text, not just in the code. Maybe:

We can tell Angular that the dependency is optional by annotating the constructor argument with `@Optional()`:

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 8]
```diff
@@ -748,105 +717,117 @@ code-example(format, language="html").
...
++ifDocsFor('ts')
+  +makeExample('dependency-injection/ts/app/providers.component.ts','import-optional', '')
++makeExample('dependency-injection/ts/app/providers.component.ts','provider-10-ctor', '')(format='.')
 
-// Skip for Dart (we have another example)
 :marked
-  Or we can provide and inject the configuration object in our top-level `AppComponent`.
-+makeExample('dependency-injection/ts/app/app.component.ts','providers', 'app/app.component.ts (providers)')(format=".")
-+makeExample('dependency-injection/ts/app/app.component.ts','ctor', 'app/app.component.ts (constructor)')(format=".")
+  Be prepared for a null logger. If we don't register one somewhere up the line,
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Instead of 2nd person here, let's continue with a "we" focus. Maybe:

When using `@Optional()`, our code must be prepared for a null value. If we don't register a logger somewhere up the line, the injector will set the value of `logger` to `null`.

File: public/docs/_examples/dependency-injection/dart/lib/logger_service.dart (hunk 6 above)
Comment (chalin) [inline public/docs/_examples/dependency-injection/dart/lib/logger_service.dart]: dartfmt doesn't seem to care. It leaves the line there if present, and doesn't add it if it is absent. I'll all for compactness, so I removed that extra line.

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 7 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: > This sounds very formal ...

Yes, what you propose is much better, thanks. Fixed.

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 8 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Fixed.

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 9]
```diff
@@ -683,61 +652,61 @@ code-example(format, language="html").
...
   In our sample, we need it only in the `HeroesComponent`,
-  where it replaces the previous `HeroService` registration in the metadata `providers` array.
+  where it replaces the previous `HeroService` registration in the metadata `providers` #{_array}.
   Here we see the new and the old implementation side-by-side:
-// #enddocregion providers-factory-5
+
+- var stylePattern = { otl: /(providers.*),$/gm };
 +makeTabs(
   `dependency-injection/ts/app/heroes/heroes.component.ts,
   dependency-injection/ts/app/heroes/heroes.component.1.ts`,
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: I get BAD FILENAME for v2 in the TS docs. (bad docregion?)

Also, the Dart docs show this:

```
import 'hero_list_component_2.dart';
import 'hero_service_1.dart';
/*
import 'hero_list_component.dart';
import 'hero_service.dart';
*/
```

They should instead look more like the v3 imports, shouldn't they?

```
import 'hero_list_component.dart';
import 'hero_service.dart';
```

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 10]
```diff
@@ -748,105 +717,117 @@ code-example(format, language="html").
...
 :marked
   We have similar good fortune when we write a constructor that requires an injected class-based dependency.
   We define a constructor parameter with the `HeroService` class type,
   and Angular knows to inject the
   service associated with that `HeroService` class token:
-// #enddocregion tokens-2
-+makeExample('dependency-injection/ts/app/providers.component.ts','provider-8-ctor')(format=".")
-// #docregion tokens-3
+
++makeExample('dependency-injection/ts/app/heroes/hero-list.component.ts', 'ctor-signature')
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: I'm so glad this doesn't say ProviderComponent8 any more!

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 9 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Good catch. Fixed.

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 10 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: :)

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 11]
```diff
@@ -1,279 +1,138 @@
...
-+includeShared('{ts}', 'tokens-opaque-2')
-+makeExample('dependency-injection/dart/lib/providers_component.dart','use-opaque-token')(format=".")
-:marked
-  Here's an example of providing configuration information
-  for an injected class. First define the class:
-+makeExample('dependency-injection/dart/lib/providers_component.dart','configurable-logger')(format=".")
-:marked
-  Then inject that class and its configuration information:
-+makeExample('dependency-injection/dart/lib/providers_component.dart','providers-7')(format=".")
+    The problem with using `Map` as a token, is not that it is an abstract
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: This is a really long sentence, and it continues the thought of the previous paragraph. Let's shorten this sentence and combine the two paragraphs. Maybe:

But what should we use as the token? We can't use `Map` because it's too general purpose. Our app might depend on several maps, each for a different purpose.

Dart difference: Interfaces are valid tokens
... are just class names. `Map` is a _valid_ token even though it's the name of an abstract class; it's just _unsuitable_ as a token because it's too general.

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 12]
```diff
@@ -748,105 +717,117 @@ code-example(format, language="html").
...
-    There is no interface type information left for Angular to find at runtime.
-// end Typescript only
-
-// #docregion tokens-opaque-1
-<a id="opaque-token"></a>
-- var lang = current.path[1]
-- var opaquetoken = lang == 'dart' ? '<code>OpaqueToken</code>' : '<a href="../api/core/OpaqueToken-class.html"><code>OpaqueToken</code></a>'
-h3 OpaqueToken
-p.
   The solution is to define and use an !{opaquetoken}.
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: By this time we might have forgotten the problem. :) How about:

> The solution to specifying non-class dependencies is...

Of course, that raises the question (especially for Dart): what should that solution be? is it the only one? Maybe "The solution" -> "One solution"?

File: public/docs/ts/latest/guide/dependency-injection.jade [hunk 13]
```diff
@@ -748,105 +717,117 @@ code-example(format, language="html").
...
+:marked
+  Now we can inject the configuration object into any constructor that needs it, with
+  the help of an `@Inject` #{_decorator}:
+
++makeExample('dependency-injection/ts/app/app.component.2.ts','ctor')(format=".")
+
+- var configType = _docsFor == 'dart' ? '<code>Map</code>' : '<code>AppConfig</code>'
 .l-sub-section
   :marked
     Although it plays no role in dependency injection,
```
Comment (kwalrath) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Define "it" before using it:

it <-> the !{configType} interface

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 14]
```diff
@@ -1,279 +1,138 @@
...
+    a custom class type:
 
-+includeShared('{ts}', 'summary')
+  +makeExample('dependency-injection/ts/app/app.config.ts','config-alt','app/app-config.ts (alternative config)')(format='.')
 
-+includeShared('{ts}', 'appendix-explicit-injector-1')
-+makeExample('dependency-injection/dart/lib/injector_component.dart', 'injector', 'lib/injector_component.dart')
-+includeShared('{ts}', 'appendix-explicit-injector-2')
+  :marked
+    One of the key benefits of defining a configuration object class, is
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Move the comma to be before "relative". Maybe change "relative..." to be clearer. I'm not sure if you're comparing the map & class benefits, or something else.

(Where the comma is, it's equivalent to "This, is that", which is you wouldn't normally do. If the sentence seems too hard to read without the comma, I'd rewrite to make it clearer. Maybe: Defining a configuration object class has a few benefits. One key benefit is strong static checking: we'll be warned early if we misspell a property name or assign it a value of the wrong type.)

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 15]
```diff
@@ -1,279 +1,138 @@
...
+    strong static checking relative to configuration property names and their types.
+    The Dart [cascade operator][cascade-op] provides a convenient means of initializing
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Cascade isn't an operator, strictly speaking. Also, be sure to show it "(`..`)", not just name it.

I'm not sure what the most important points of this discussion are. It needs to be clearer, and probably broken into multiple paragraphs.

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 16]
```diff
@@ -1,279 +1,138 @@
...
+    the configuration object. Use of the cascaded object initialization idiom
+    prevents us from defining our configuration object instance as `const`. 
+    As a consequencem we cannot declare a [value provider](#value-provider),
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: consequencem -> consequence,

(if it isn't rewritten out of existence)

**Comment by kwalrath** [comment]
Looks great! The only real problem I saw was the "key benefits of defining a configuration object class" paragraph, which needs to be clarified and probably broken up.

**Comment by kwalrath** [comment]
@Foxandxss & @wardbell: I don't know the TS side, so you still might want to take a look to make sure the new text is OK by you...

**Comment by wardbell** [comment]
Will look

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 11 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: fixed per our discussion

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 12 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: Both good points. Fixed (and yes, it is only one solution).

File: public/docs/ts/latest/guide/dependency-injection.jade (hunk 13 above)
Comment (chalin) [inline public/docs/ts/latest/guide/dependency-injection.jade]: done

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 14 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: I prefer your rewrite. Fixed.

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 15 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Fixed. Changed 'operator' to 'syntax'.

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 16 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: fixed

**Comment by chalin** [comment]
Updates done.

> Looks great! The only real problem I saw was the "key benefits..." paragraph

That paragraph has been reworked.

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 17]
```diff
@@ -107,32 +109,32 @@ block what-should-we-use-as-token
...
-    strong static checking relative to configuration property names and their types.
-    The Dart [cascade operator][cascade-op] provides a convenient means of initializing
-    the configuration object. Use of the cascaded object initialization idiom
+    Defining a configuration class has a few benefits. One key benefit
+    is strong static checking: we'll be warned early if we misspell a property
+    name or assign it a value of the wrong type.
+    The Dart [cascade notation][cascade] (`..`) provides a convenient means of initializing
+    a configuration object. 
+
+    Use of the cascaded initialization idiom
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: On the right track, but this could still be simplified. Maybe:

If we use cascades, the configuration object can't be declared `const` and we can't use
a [value provider](...). The solution is to use a [factory provider](...):

File: public/docs/dart/latest/guide/dependency-injection.jade [hunk 18]
```diff
@@ -107,32 +109,32 @@ block what-should-we-use-as-token
...
     prevents us from defining our configuration object instance as `const`. 
-    As a consequencem we cannot declare a [value provider](#value-provider),
-    but a [factory provider](#factory-provider) does the trick &mdash; as
-    we illustrate by providing and injecting the configuration object in our
+    As a consequence, we cannot declare a [value provider](#value-provider),
+    but a [factory provider](#factory-provider) does the trick.
+
+    We illustrate this below by providing and injecting the configuration object in our
```
Comment (kwalrath) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Could we just delete this paragraph? If not, let's simplify the wording.

**Comment by kwalrath** [comment]
I had a [couple more quibbles about the cascade wording](https://github.com/angular/angular.io/pull/1573/commits/dbcff2ddbbf88033ec3cb730c6b2bb66d49fda91#r65644008), but otherwise LGTM.

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 17 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Yes, what you propose is better. Done.

File: public/docs/dart/latest/guide/dependency-injection.jade (hunk 18 above)
Comment (chalin) [inline public/docs/dart/latest/guide/dependency-injection.jade]: Actually, that paragraph is there to match up with the TS prose which reads:
> ... we can provide and inject the configuration object in our top-level `AppComponent`.

How about:
> We illustrate this next. We also show how to provide and inject the
    configuration object in our top-level `AppComponent`:

**Comment by chalin** [comment]
Updates done.

**Comment by kwalrath** [comment]
LGTM. I'll merge this unless someone screams NO...
